### PR TITLE
Optimize raw message body encoding

### PR DESCRIPTION
**Pull request by pentp**
* Don't encode PooledResponse<T> wrapper types for normal response messages, only encode T.

Fixed the issues that caused #8093.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8106)

### CODE REVIEW

File: src/Orleans.Core/Networking/Connection.cs [hunk 1]
```diff
@@ -105,13 +105,13 @@ private static Task OnConnectedAsync(ConnectionContext context)
...
 
-        protected virtual async Task RunInternal()
+        protected virtual Task RunInternal()
         {
             _transport = this.Context.Transport;
             _processIncomingTask = this.ProcessIncoming();
             _processOutgoingTask = this.ProcessOutgoing();
             _initializationTcs.TrySetResult(0);
-            await Task.WhenAll(_processIncomingTask, _processOutgoingTask);
+            return Task.WhenAll(_processIncomingTask, _processOutgoingTask);
```
Comment (ReubenBond) [inline src/Orleans.Core/Networking/Connection.cs:114]: This was written this way for debugging purposes. There are generally only 10s-100s of connections, and they are each very long-lived, so per-connection state machine allocations are ok.

File: src/Orleans.Serialization/Invocation/Response.cs [hunk 2]
```diff
@@ -170,6 +177,101 @@ public override void Dispose()
...
+    {
+        public abstract void WriteRaw<TBufferWriter>(ref Writer<TBufferWriter> writer, object value) where TBufferWriter : IBufferWriter<byte>;
+        public abstract object ReadRaw<TInput>(ref Reader<TInput> reader, scoped ref Field field);
+    }
+
+    [RegisterSerializer]
+    internal sealed class PooledResponseCodec<TResult> : ResponseCodec, IFieldCodec<Response<TResult>>
+    {
+        private readonly Type _codecFieldType = typeof(Response<TResult>);
+        private readonly Type _resultType = typeof(TResult);
```
Comment (ReubenBond) [inline src/Orleans.Serialization/Invocation/Response.cs:193]: Is there a reason to not make these static?

Comment (pentp) [reply src/Orleans.Serialization/Invocation/Response.cs:193]: Accessing static fields on generic types with reference type parameters is relatively slow (through a dictionary).

File: src/Orleans.Core/Messaging/MessageSerializer.cs [hunk 3]
```diff
@@ -131,13 +128,42 @@ private void ReadBodyObject<TInput>(Message message, ref Reader<TInput> reader)
         {
             var field = reader.ReadFieldHeader();
 
-            var bodyCodec = _codecProvider.GetCodec(field.FieldType);
-            message.BodyObject = bodyCodec.ReadValue(ref reader, field);
+            if (message.Result == ResponseTypes.Success)
+            {
+                message.Result = ResponseTypes.None; // reset raw response indicator
```
Comment (ReubenBond) [inline src/Orleans.Core/Messaging/MessageSerializer.cs:133]: I'll merge this as-is. In a subsequent PR, I'd like us to have some more explanatory comments about how the ResponseTypes affect body encoding.

File: src/Orleans.Core/Messaging/MessageSerializer.cs [hunk 4]
```diff
@@ -131,13 +128,42 @@ private void ReadBodyObject<TInput>(Message message, ref Reader<TInput> reader)
...
             var headers = message.Headers;
+            IFieldCodec? bodyCodec = null;
+            ResponseCodec? rawCodec = null;
+            if (message.BodyObject is not null)
+            {
+                bodyCodec = _codecProvider.GetCodec(message.BodyObject.GetType());
+                if (headers.ResponseType is ResponseTypes.None && bodyCodec is ResponseCodec responseCodec)
+                {
+                    rawCodec = responseCodec;
+                    headers.ResponseType = ResponseTypes.Success; // indicates a raw simple response (not wrapped in Response<T>)
```
Comment (ReubenBond) [inline src/Orleans.Core/Messaging/MessageSerializer.cs:163]: Ideally, we should reset this before returning to the caller.

**Review by ReubenBond — APPROVED** [review]

**Comment by ReubenBond** [comment]
I believe it would be cleaner to have a `BodyEncoding` enum stuffed into the headers, separate from the `ResponseTypes`

**Pull request merged**